### PR TITLE
fix(sdk): accept fractional durations in normalize_time_string

### LINKED ISSUE #14441: feat(sdk): have normalize_time_string support fractional durations
What happened?

normalize_time_string (called by convert_duration_to_seconds in sdk/python/kfp/dsl/structures.py, lines ~1022–1046) rejects valid fractional duration strings such as 0.5h or 1.5m, even though Kubernetes supports decimal quantities for durations and resources.

The function splits the input on non-digit characters using re.split(r'(\D+)', no_ws_duration). For an integer duration like 1h, this correctly produces two elements (['1', 'h']). But for a fractional duration like 0.5h, the decimal point is itself matched by \D+, so the string splits into four elements: ['0', '.', '5', 'h']. Since the function requires exactly two elements, it raises ValueError instead of parsing the value.

This blocks valid pipeline configurations that rely on fractional durations, e.g.:

task.set_retry(num_retries=2, backoff_duration='0.5h')

What did you expect to happen?

Fractional durations should parse correctly and convert to the right number of seconds, e.g. '0.5h' → 1800, '1.5m' → 90.

Steps to reproduce

from kfp.dsl.structures import convert_duration_to_seconds

cases = ['1h', '30m', '0.5h', '1.5m', '2.5s', '0.25h']
for c in cases:
    try:
        print(f"{c:>8} -> {convert_duration_to_seconds(c)}")
    except ValueError as e:
        print(f"{c:>8} -> ValueError: {e}")

Output:

      1h -> 3600
     30m -> 1800
    0.5h -> ValueError: Invalid duration string: '0.5h'. Expected one value (as integer in string) and one unit, such as '1 hour'.
    1.5m -> ValueError: Invalid duration string: '1.5m'. Expected one value (as integer in string) and one unit, such as '1 hour'.
    2.5s -> ValueError: Invalid duration string: '2.5s'. Expected one value (as integer in string) and one unit, such as '1 hour'.
   0.25h -> ValueError: Invalid duration string: '0.25h'. Expected one value (as integer in string) and one unit, such as '1 hour'.

Environment

KFP SDK: current master
Python: 3.14
OS: Ubuntu

Additional context

I'm working on a fix (regex-based parsing anchored with re.fullmatch, plus a check that the resulting seconds value is a whole number) and will open a PR referencing this issue.

### PR DESCRIPTION
**Pull request by ravikumar266**
Fixes #14441

normalize_time_string was rejecting valid fractional durations like 0.5h or 1.5m. The root cause: it split the input on non-digit characters (re.split(r'(\D+)', ...)), and a decimal point counts as non-digit too — so 0.5h split into ['0', '.', '5', 'h'] instead of the expected two-part [value, unit], which tripped a length check and raised ValueError.

This matters because Kubernetes accepts decimal durations, so something like RetryPolicy.backoff_duration='0.5h' should be valid but currently isn't.

What I changed:

normalize_time_string now parses with one anchored regex (\d+(?:\.\d+)?[a-zA-Z]+) instead of splitting on non-digits.
convert_duration_to_seconds computes the seconds as a float now, and raises a clear error if the result isn't a whole number (e.g. 0.5s) instead of crashing on int('0.5').

Before:

1h    -> 3600
30m   -> 1800
0.5h  -> ValueError: Invalid duration string: '0.5h'. Expected one value...
1.5m  -> ValueError: ...
2.5s  -> ValueError: ...
0.25h -> ValueError: ...

After:

0.5h  -> 1800
1.5m  -> 90
0.25h -> 900
0.5s  -> ValueError: ...must resolve to a whole number of seconds, but got 0.5.
1.h, .5h -> still rejected (malformed, correctly)

Added test cases to the existing TestNormalizeTimeString and TestConvertDurationToSeconds classes, plus two new tests for the edge cases: malformed decimals should still fail, and durations that don't resolve to whole seconds should fail with a clearer message than before.

One note for reviewers — I ran the full sdk/python/kfp/dsl suite and got 4 unrelated failures. I checked these aren't caused by this PR by stashing my changes and running against clean master, same failures show up:

2 tests fail on ModuleNotFoundError: nbformat — just a missing optional dep on my machine
2 tests in custom_artifact_types_test.py fail with AttributeError: 'Constant' object has no attribute 's' — looks like a pre-existing incompatibility with Python 3.14's AST module

### CODE REVIEW

File: sdk/python/kfp/dsl/structures.py [hunk 1]
```diff
@@ -1065,7 +1066,12 @@ def convert_duration_to_seconds(duration: str) -> int:
     if duration[-1] not in seconds_per_unit.keys():
         raise ValueError(
             f"Unsupported duration unit: '{duration[-1]}' for '{duration}'.")
-    return int(duration[:-1]) * seconds_per_unit[duration[-1]]
+    seconds = float(duration[:-1]) * seconds_per_unit[duration[-1]]
+    if not seconds.is_integer():
```
[the remark this answers is not in the document]
Comment (ravikumar266) [reply sdk/python/kfp/dsl/structures.py]: Good catch — switched to Decimal for exact arithmetic instead of binary float, and added a regression test for 2.05m specifically (previously would have raised under the old float-based check). Pushed in 2fa86b174.

File: sdk/RELEASE.md [hunk 2]
```diff
@@ -14,6 +14,8 @@
 
 ## Bug fixes and other changes
 
+* Fix `normalize_time_string`/`convert_duration_to_seconds` to accept fractional durations (e.g. `'0.5h'`, `'1.5m'`), which previously raised `ValueError` due to an incorrect string-splitting check.
+
```
Comment (zazulam) [inline sdk/RELEASE.md]: Please discard this change as the release gets updated in the github releases page

**Review by zazulam — CHANGES_REQUESTED** [review]
1 small comment

File: sdk/RELEASE.md (hunk 2 above)
Comment (ravikumar266) [reply sdk/RELEASE.md]: Done, removed the RELEASE.md change. Thanks!

**Comment by ravikumar266** [comment]
Hi @zazulam, thanks for the review. I've removed the `sdk/RELEASE.md` change as requested, and all CI checks are passing now.

Could you please take another look? If it looks good, could you add `/lgtm` and `/approve` so Tide can merge it?

Thanks for your time!
